tests/pulumirpc/log.go

```diff
@@ -0,0 +1,245 @@
+	Metadata interface{} `json:"metadata,omitempty"`
+}
+
+type RegisterResourceList []RegisterResource
+
+// Named returns RegisterResource entries matching the given resource type, name, and parent.
+func (l RegisterResourceList) Named(parent resource.URN, typ tokens.Type, name tokens.QName) RegisterResourceList {
+	var results []RegisterResource
+	for _, v := range l {
+		if resource.URN(v.Request.Parent) == parent && tokens.Type(v.Request.Type) == typ && tokens.QName(v.Request.Name) == name {
+			results = append(results, v)
+		}
+	}
+	return results
+}
+
+// ParseRegisterResource parses a log entry as a RegisterResource call.
+func ParseRegisterResource(entry DebugInterceptorLogEntry) (*RegisterResource, bool) {
+	if entry.Method != "/pulumirpc.ResourceMonitor/RegisterResource" {
+		return nil, false
+	}
+	request := &pulumirpc.RegisterResourceRequest{}
+	err := protojson.Unmarshal(entry.Request, request)
+	if err != nil {
+		return nil, false
+	}
+	response := &pulumirpc.RegisterResourceResponse{}
+	err = protojson.Unmarshal(entry.Response, response)
+	if err != nil {
+		return nil, false
+	}
+	result := RegisterResource{
+		Request:  *request,
+		Response: *response,
+		Errors:   entry.Errors,
+		Metadata: entry.Metadata,
+	}
+	return &result, true
+}
+
+// Invokes lists the Invoke calls in the log.
+func (l DebugInterceptorLogEntryList) Invokes() InvokeList {
+	var results []Invoke
+	for _, entry := range l {
+		if r, ok := ParseInvoke(entry); ok {
+			results = append(results, *r)
+		}
+	}
+	return results
+}
+
+// Invoke is a decoded "/pulumirpc.ResourceMonitor/Invoke" RPC call.
+type Invoke struct {
+	Request  pulumirpc.ResourceInvokeRequest
+	Response pulumirpc.InvokeResponse
+	Errors   []string    `json:"errors,omitempty"`
+	Metadata interface{} `json:"metadata,omitempty"`
+}
+
+type InvokeList []Invoke
+
+// Lookup returns Invoke entries matching the given token (operation).
+func (l InvokeList) Tok(tok string) InvokeList {
+	var results []Invoke
+	for _, v := range l {
+		if v.Request.Tok == tok {
+			results = append(results, v)
+		}
+	}
+	return results
+}
+
+// Lookup returns Invoke entries matching the given provider and token (operation).
+func (l InvokeList) ByProvider(providerUrn resource.URN) InvokeList {
+	var results []Invoke
+	for _, v := range l {
+		if resource.URN(v.Request.Provider) == providerUrn {
+			results = append(results, v)
+		}
+	}
+	return results
+}
+
+// ParseInvoke parses a log entry as a Invoke call.
+func ParseInvoke(entry DebugInterceptorLogEntry) (*Invoke, bool) {
+	if entry.Method != "/pulumirpc.ResourceMonitor/Invoke" {
+		return nil, false
+	}
+	request := &pulumirpc.ResourceInvokeRequest{}
+	err := protojson.Unmarshal(entry.Request, request)
+	if err != nil {
+		return nil, false
+	}
+	response := &pulumirpc.InvokeResponse{}
+	err = protojson.Unmarshal(entry.Response, response)
+	if err != nil {
+		return nil, false
+	}
+	result := Invoke{
+		Request:  *request,
```

Comment:
nit: I usually prefer to declare the struct itself, so we only need to reference it in the unmarshal functions, eg.
```go
var request pulumirpc.ResourceInvokeRequest{}
protojson.Unmarshal(entry.Request, &request)
result := Invoke{
		Request:  request,
...
```